File: src/main/java/mods/eln/sim/mna/component/InterSystemAbstraction.java
package mods.eln.sim.mna.component;

import mods.eln.sim.mna.RootSystem;
import mods.eln.sim.mna.SubSystem;
import mods.eln.sim.mna.misc.IDestructor;
import mods.eln.sim.mna.misc.IRootSystemPreStepProcess;
import mods.eln.sim.mna.state.State;
import mods.eln.sim.mna.state.VoltageState;
import mods.eln.sim.mna.SubSystem.Thevenin;

public class InterSystemAbstraction implements IAbstractor, IDestructor, IRootSystemPreStepProcess {

    VoltageState aNewState;
    Resistor aNewResistor;
    VoltageSource aNewDelay;
    VoltageState bNewState;
    Resistor bNewResistor;
    VoltageSource bNewDelay;

    RootSystem root;
    Resistor interSystemResistor;

    State aState;
    State bState;
    SubSystem aSystem;
    SubSystem bSystem;

    public InterSystemAbstraction(RootSystem root, Resistor interSystemResistor) {
        this.interSystemResistor = interSystemResistor;
        this.root = root;

        aState = interSystemResistor.aPin;
        bState = interSystemResistor.bPin;
        aSystem = aState.getSubSystem();
        bSystem = bState.getSubSystem();

        aSystem.interSystemConnectivity.add(bSystem);
        bSystem.interSystemConnectivity.add(aSystem);

        aNewState = new VoltageState();
        aNewResistor = new Resistor();
        aNewDelay = new VoltageSource("aNewDelay");
        bNewState = new VoltageState();
        bNewResistor = new Resistor();
        bNewDelay = new VoltageSource("bNewDelay");

        aNewResistor.connectGhostTo(aState, aNewState);
        aNewDelay.connectTo(aNewState, null);
        bNewResistor.connectGhostTo(bState, bNewState);
        bNewDelay.connectTo(bNewState, null);

        calibrate();

        aSystem.addComponent(aNewResistor);
        aSystem.addState(aNewState);
        aSystem.addComponent(aNewDelay);
        bSystem.addComponent(bNewResistor);
        bSystem.addState(bNewState);
        bSystem.addComponent(bNewDelay);

        aSystem.breakDestructor.add(this);
        bSystem.breakDestructor.add(this);

        interSystemResistor.abstractedBy = this;

        root.addProcess(this);
    }

    void calibrate() {
        double voltage = (aState.state + bState.state) / 2;
        aNewDelay.setVoltage(voltage);
        bNewDelay.setVoltage(voltage);

        double resistance = interSystemResistor.getResistance() / 2;
        aNewResistor.setResistance(resistance);
        bNewResistor.setResistance(resistance);
    }

    @Override
    public void dirty(Component component) {
        calibrate();
    }

    @Override
    public SubSystem getAbstractorSubSystem() {
        return aSystem;
    }

    @Override
    public void destruct() {
        aSystem.breakDestructor.remove(this);
        aSystem.removeComponent(aNewDelay);
        aSystem.removeComponent(aNewResistor);
        aSystem.removeState(aNewState);
        bSystem.breakDestructor.remove(this);
        bSystem.removeComponent(bNewDelay);
        bSystem.removeComponent(bNewResistor);
        bSystem.removeState(bNewState);

        root.removeProcess(this);

        interSystemResistor.abstractedBy = null;

        aSystem.component.add(interSystemResistor);
    }

    @Override
    public void rootSystemPreStepProcess() {
        Thevenin a = aNewDelay.getSubSystem().getTh(aState,aNewDelay);
        Thevenin b = bNewDelay.getSubSystem().getTh(bState,bNewDelay);

        double voltage = (a.voltage - b.voltage) * b.resistance / (a.resistance + b.resistance) + b.voltage;
        if (Double.isNaN(voltage)) {
            voltage = 0;
        }

        aNewDelay.setVoltage(voltage);
        bNewDelay.setVoltage(voltage);
    }
}

File: src/main/java/mods/eln/sim/mna/active/InterSystemAbstraction.kt
package mods.eln.sim.mna.active

import mods.eln.sim.mna.RootSystem
import mods.eln.sim.mna.SubSystem
import mods.eln.sim.mna.passive.Component
import mods.eln.sim.mna.passive.Resistor
import mods.eln.sim.mna.misc.IAbstractor
import mods.eln.sim.mna.misc.IDestructor
import mods.eln.sim.mna.state.State
import mods.eln.sim.mna.state.VoltageState

class InterSystemAbstraction(internal var root: RootSystem, internal var interSystemResistor: Resistor) : IAbstractor, IDestructor {

    internal var aNewState: VoltageState
    internal var aNewResistor: Resistor
    internal var aNewDelay: DelayInterSystem2
    internal var bNewState: VoltageState
    internal var bNewResistor: Resistor
    internal var bNewDelay: DelayInterSystem2
    internal var thevnaCalc: DelayInterSystem2.ThevnaCalculator

    internal var aState: State = interSystemResistor.aPin ?: throw Exception("aPin on InterSystemResistor cannot be null!")
    internal var bState: State = interSystemResistor.bPin ?: throw Exception("bPin on InterSystemResistor cannot be null!")
    override var abstractorSubSystem: SubSystem = aState.subSystem ?: throw Exception("subsystem connected to aPin cannot be null!")
    internal var bSystem: SubSystem = bState.subSystem ?: throw Exception("subsystem connected to bPin cannot be null!")

    init {
        abstractorSubSystem.interSystemConnectivity.add(bSystem)
        bSystem.interSystemConnectivity.add(abstractorSubSystem)

        aNewState = VoltageState()
        aNewResistor = Resistor()
        aNewDelay = DelayInterSystem2()
        bNewState = VoltageState()
        bNewResistor = Resistor()
        bNewDelay = DelayInterSystem2()

        aNewResistor.connectGhostTo(aState, aNewState)
        aNewDelay.connectTo(aNewState, null)
        bNewResistor.connectGhostTo(bState, bNewState)
        bNewDelay.connectTo(bNewState, null)

        calibrate()

        abstractorSubSystem.addComponent(aNewResistor)
        abstractorSubSystem.addState(aNewState)
        abstractorSubSystem.addComponent(aNewDelay)
        bSystem.addComponent(bNewResistor)
        bSystem.addState(bNewState)
        bSystem.addComponent(bNewDelay)

        abstractorSubSystem.breakDestructor.add(this)
        bSystem.breakDestructor.add(this)

        interSystemResistor.abstractedBy = this

        thevnaCalc = DelayInterSystem2.ThevnaCalculator(aNewDelay, bNewDelay)
        root.addProcess(thevnaCalc)
    }

    internal fun calibrate() {
        val u = (aState.state + bState.state) / 2
        aNewDelay.u = u
        bNewDelay.u = u

        val r = interSystemResistor.r / 2
        aNewResistor.r = r
        bNewResistor.r = r
    }

    override fun dirty(component: Component) {
        calibrate()
    }

    override fun destruct() {
        abstractorSubSystem.breakDestructor.remove(this)
        abstractorSubSystem.removeComponent(aNewDelay)
        abstractorSubSystem.removeComponent(aNewResistor)
        abstractorSubSystem.removeState(aNewState)
        bSystem.breakDestructor.remove(this)
        bSystem.removeComponent(bNewDelay)
        bSystem.removeComponent(bNewResistor)
        bSystem.removeState(bNewState)

        root.removeProcess(thevnaCalc)

        interSystemResistor.abstractedBy = null

        abstractorSubSystem.component.add(interSystemResistor)
    }
}
